STYLE.md

```diff
@@ -0,0 +1,317 @@
+# Vector Style Guide
+
+As a large, open-source project, it can be a struggle to ensure a consistent level of code quality.
+This style guide is meant to be the canonical reference for all things style: code comments,
+formatting, acceptable (or unacceptable) crates, data structures, or algorithms, and so on.
+
+In essence, we hope to turn pull request review comments like "why did you do it this way?" or "I
+think you could try doing it this way" into "we always do X this way: <link to style guide>".
+
+# Formatting
+
+At a high-level, code formatting is straightforward: we use the native `rustfmt` exclusively, and
+comprehensively.  All Rust source code within the repository should be formatted using `rustfmt`.
+
+You can acquire `rustfmt` -- which is invoked as `cargo fmt` -- by following the directions listed
+out [in the rustfmt repository](https://github.com/rust-lang/rustfmt#on-the-stable-toolchain).
+
+Vector has its own formatting rules (`.rustfmt.toml`) that will automatically be used when you run
+`cargo fmt` within the repository.  If you're ever in doubt, you can also run `make check-fmt` which
+will invoke `cargo fmt` in a dry-run mode, checking to see if any changed files are not currently
+formatted correctly.
+
+As an additional note, `rustfmt` sometimes can fail to format code within macros, so if you happen
+to see such code that doesn't look like it's formatted correctly, you may need to manually tweak it
+if `rustfmt` cannot be persuaded to format it correctly for you. :)
+
```

Comment:
Could we add a section on code documentation as per here: https://github.com/vectordotdev/vector/blob/master/docs/REVIEWING.md?plain=1#L76-L89

Reply:
Created #14573 to track doing so as a follow-up item.